src/dialogs/editor.tsx

```diff
@@ -0,0 +1,258 @@
+        this.state = new EditorState();
+
+        this.load_file();
+
+        cockpit.spawn(['test', '-w', filename], { superuser: "try" })
+                .then(() => this.update({ writable: true }))
+                .catch(() => this.update({ writable: false }));
+
+        const handle_watch = debounce(500, (_content: string | null, tag_now: string | null) => {
+            this.update({ tag_now });
+        });
+
+        this.file.watch(handle_watch, { read: false });
+    }
+
+    async save() {
+        if (!this.state.tag_now) {
+            console.error("Unable to save as 'tag_now' is not initialised");
+            return;
+        }
+
+        try {
+            this.update({ saving: true });
+            console.log("save", this.state.content, this.state.tag_now);
+            // Force the tag to be "-" when the file is not found to allow the user to write the file.
+            const tag = await this.file.replace(this.state.content, this.state.error?.problem === "not-found"
+                ? "-"
+                : this.state.tag_now);
+            this.update({ saving: false, modified: false, tag_now: tag, tag_at_load: tag });
+        } catch (exc: any /* eslint-disable-line @typescript-eslint/no-explicit-any */) {
+            this.update({ error: exc, saving: false });
+        }
+    }
+
+    close() {
+        this.file.close();
+    }
+}
+
+export const EditFileModal = ({ dialogResult, path } : {
+    dialogResult: DialogResult<void>,
+    path: string
+}) => {
+    const [last_tag, setLastTag] = React.useState<string | null>(null);
+    const [state, setState] = React.useState(new EditorState());
+    const [editor, setEditor] = React.useState<Editor | null>(null);
+
+    React.useEffect(() => {
+        const editor = new Editor(path);
+        editor.on('updated', setState);
+        setEditor(editor);
+        return () => {
+            editor.close();
+        };
+    }, [path]);
+
+    const { modified } = state;
+    React.useEffect(() => {
+        const before_unload = (event: BeforeUnloadEvent) => {
+            event.preventDefault();
+
+            // Included for legacy support, e.g. Chrome/Edge < 119
+            event.returnValue = true;
+        };
+        if (modified) {
+            window.addEventListener('beforeunload', before_unload);
+            return () => {
+                window.removeEventListener('beforeunload', before_unload);
+            };
+        }
+    }, [modified]);
+
+    const handleEscape = (event: KeyboardEvent) => {
+        if (state?.modified) {
+            event.preventDefault();
+        } else {
+            dialogResult.resolve();
+        }
+    };
+
+    /* Translators: $0 represents a filename */
+    let title = <>{cockpit.format(state?.writable ? _("Edit “$0”") : _("View “$0”"), path)}</>;
+    if (!state.writable) {
+        // TODO: dark mode and lack of spacing
```

Comment:
Lack of spacing

![image](https://github.com/user-attachments/assets/7db95e23-7e92-4edb-8b96-5d1478dae853)

Dark mode:

![image](https://github.com/user-attachments/assets/bdb97b37-7576-4a1e-bde6-d9013fe2da10)

Reply:
TODO: can go now :)